Funky path causing PSES crash

I've been editing the PowerShell build.psm1 file (PowerShell folder open in VSCode) and I've been noticing PSES crash after some period of time.  

This is on VSCode 1.17.2 x64, PSES 1.5.0 on Windows 10 FCU.

Also, I might have been editing in one of the Git diff windows.

Here's the relevant part of the log:
```
11/4/2017 5:29:46 PM [VERBOSE] - Method "ReadMessage" at line 123 of C:\projects\powershelleditorservices\src\PowerShellEditorServices.Protocol\MessageProtocol\MessageReader.cs

    Received Request 'textDocument/documentHighlight' with id 221

11/4/2017 5:29:46 PM [VERBOSE] - Method "ResolveFilePath" at line 354 of C:\projects\powershelleditorservices\src\PowerShellEditorServices\Workspace\Workspace.cs

    Resolved path: git:/c%3A/Users/Keith/GitHub/rkeithhill/PowerShell/build.psm1?%7B%22path%22%3A%22c%3A%5C%5CUsers%5C%5CKeith%5C%5CGitHub%5C%5Crkeithhill%5C%5CPowerShell%5C%5Cbuild.psm1%22%2C%22ref%22%3A%22~%22%7D

11/4/2017 5:29:46 PM [VERBOSE] - Method "WriteMessage" at line 61 of C:\projects\powershelleditorservices\src\PowerShellEditorServices.Protocol\MessageProtocol\MessageWriter.cs

    Writing Response 'textDocument/documentHighlight' with id 221

11/4/2017 5:29:47 PM [VERBOSE] - Method "ReadMessage" at line 123 of C:\projects\powershelleditorservices\src\PowerShellEditorServices.Protocol\MessageProtocol\MessageReader.cs

    Received Request 'textDocument/definition' with id 222

11/4/2017 5:29:47 PM [VERBOSE] - Method "ResolveFilePath" at line 354 of C:\projects\powershelleditorservices\src\PowerShellEditorServices\Workspace\Workspace.cs

    Resolved path: git:/c%3A/Users/Keith/GitHub/rkeithhill/PowerShell/build.psm1?%7B%22path%22%3A%22c%3A%5C%5CUsers%5C%5CKeith%5C%5CGitHub%5C%5Crkeithhill%5C%5CPowerShell%5C%5Cbuild.psm1%22%2C%22ref%22%3A%22~%22%7D

11/4/2017 5:29:47 PM [ERROR] - Method "OnListenTaskCompleted" at line 391 of C:\projects\powershelleditorservices\src\PowerShellEditorServices.Protocol\MessageProtocol\ProtocolEndpoint.cs

    ProtocolEndpoint message loop terminated due to unhandled exception:
    
    System.AggregateException: One or more errors occurred. ---> System.UriFormatException: Invalid URI: The hostname could not be parsed.
       at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
       at Microsoft.PowerShell.EditorServices.Protocol.Server.LanguageServer.GetFileUri(String filePath)
       at Microsoft.PowerShell.EditorServices.Protocol.Server.LanguageServer.<HandleDefinitionRequest>d__33.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at Microsoft.PowerShell.EditorServices.Protocol.MessageProtocol.MessageDispatcher.<DispatchMessage>d__7.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at Microsoft.PowerShell.EditorServices.Protocol.MessageProtocol.ProtocolEndpoint.<ListenForMessages>d__36.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at Microsoft.PowerShell.EditorServices.Utility.AsyncContext.Start(Func`1 asyncMainFunc, ILogger logger)
       at System.Threading.Tasks.Task.Execute()
       --- End of inner exception stack trace ---
    ---> (Inner Exception #0) System.UriFormatException: Invalid URI: The hostname could not be parsed.
       at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
       at Microsoft.PowerShell.EditorServices.Protocol.Server.LanguageServer.GetFileUri(String filePath)
       at Microsoft.PowerShell.EditorServices.Protocol.Server.LanguageServer.<HandleDefinitionRequest>d__33.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at Microsoft.PowerShell.EditorServices.Protocol.MessageProtocol.MessageDispatcher.<DispatchMessage>d__7.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at Microsoft.PowerShell.EditorServices.Protocol.MessageProtocol.ProtocolEndpoint.<ListenForMessages>d__36.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at Microsoft.PowerShell.EditorServices.Utility.AsyncContext.Start(Func`1 asyncMainFunc, ILogger logger)
       at System.Threading.Tasks.Task.Execute()<---
``

--- a/src/PowerShellEditorServices/Workspace/Workspace.cs
+++ b/src/PowerShellEditorServices/Workspace/Workspace.cs
@@ -390,22 +390,26 @@ namespace Microsoft.PowerShell.EditorServices
             // view of the current file or an untitled file.
             try
             {
-                // NotSupportedException or ArgumentException gets thrown when
-                // given an invalid path.  Since any non-standard path will
-                // trigger this, assume that it means it's an in-memory file
-                // unless the path starts with file:
-                Path.GetFullPath(filePath);
+                // File system absoulute paths will have a URI scheme of file:.
+                // Other schemes like "untitled:" and "gitlens-git:" will return false for IsFile.
+                var uri = new Uri(filePath);
+                isInMemory = !uri.IsFile;
             }
-            catch (ArgumentException)
+            catch (UriFormatException)
             {
-                isInMemory = true;
-            }
-            catch (NotSupportedException)
-            {
-                isInMemory = true;
+                // Relative file paths cause a UriFormatException.
+                // In this case, fallback to using Path.GetFullPath().
+                try
+                {
+                    Path.GetFullPath(filePath);
+                }
+                catch (Exception ex) when (ex is ArgumentException || ex is NotSupportedException)
+                {
+                    isInMemory = true;
+                }
             }
 
-            return !filePath.ToLower().StartsWith("file:") && isInMemory;
+            return isInMemory;
         }
 
         private string GetBaseFilePath(string filePath)

--- a/src/PowerShellEditorServices/Workspace/Workspace.cs
+++ b/src/PowerShellEditorServices/Workspace/Workspace.cs
@@ -407,6 +407,10 @@ namespace Microsoft.PowerShell.EditorServices
                 {
                     isInMemory = true;
                 }
+                catch (PathTooLongException)
+                {
+                    // If we ever get here, it should be an actual file so, not in memory
+                }
             }
 
             return isInMemory;

--- a/test/PowerShellEditorServices.Test/Session/WorkspaceTests.cs
+++ b/test/PowerShellEditorServices.Test/Session/WorkspaceTests.cs
@@ -35,5 +35,35 @@ namespace Microsoft.PowerShell.EditorServices.Test.Session
             Assert.Equal(@"..\PeerPath\FilePath.ps1", workspace.GetRelativePath(testPathOutside));
             Assert.Equal(testPathAnotherDrive, workspace.GetRelativePath(testPathAnotherDrive));
         }
+
+        [Fact]
+        public void CanDetermineIsPathInMemory()
+        {
+            var tempDir = Environment.GetEnvironmentVariable("TEMP");
+
+            var shortDirPath = Path.Combine(tempDir, "GitHub", "PowerShellEditorServices");
+            var shortFilePath = Path.Combine(shortDirPath, "foo.ps1");
+            var shortDirUriForm = new Uri(shortDirPath).ToString();
+            var shortFileUriForm = new Uri(shortFilePath).ToString();
+
+            // Test short file absolute paths
+            Assert.False(Workspace.IsPathInMemory(shortDirPath));
+            Assert.False(Workspace.IsPathInMemory(shortFilePath));
+            Assert.False(Workspace.IsPathInMemory(shortDirUriForm));
+            Assert.False(Workspace.IsPathInMemory(shortFileUriForm));
+
+            // Test short file relative paths - not sure we'll ever get these but just in case
+            Assert.False(Workspace.IsPathInMemory("foo.ps1"));
+            Assert.False(Workspace.IsPathInMemory(".." + Path.DirectorySeparatorChar + "foo.ps1"));
+
+            // Test short non-file paths
+            Assert.True(Workspace.IsPathInMemory("untitled:untitled-1"));
+            var shortUriForm = "git:/c%3A/Users/Keith/GitHub/dahlbyk/posh-git/src/PoshGitTypes.ps1?%7B%22path%22%3A%22c%3A%5C%5CUsers%5C%5CKeith%5C%5CGitHub%5C%5Cdahlbyk%5C%5Cposh-git%5C%5Csrc%5C%5CPoshGitTypes.ps1%22%2C%22ref%22%3A%22~%22%7D";
+            Assert.True(Workspace.IsPathInMemory(shortUriForm));
+
+            // Test long non-file path - known to have crashed PSES
+            var longUriForm = "gitlens-git:c%3A%5CUsers%5CKeith%5CGitHub%5Cdahlbyk%5Cposh-git%5Csrc%5CPoshGitTypes%3Ae0022701.ps1?%7B%22fileName%22%3A%22src%2FPoshGitTypes.ps1%22%2C%22repoPath%22%3A%22c%3A%2FUsers%2FKeith%2FGitHub%2Fdahlbyk%2Fposh-git%22%2C%22sha%22%3A%22e0022701fa12e0bc22d0458673d6443c942b974a%22%7D";
+            Assert.True(Workspace.IsPathInMemory(longUriForm));
+        }
     }
 }

--- a/test/PowerShellEditorServices.Test/Session/WorkspaceTests.cs
+++ b/test/PowerShellEditorServices.Test/Session/WorkspaceTests.cs
@@ -3,12 +3,10 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 //
 
-using Microsoft.PowerShell.EditorServices;
 using System;
 using System.IO;
-using System.Linq;
-using Xunit;
 using Microsoft.PowerShell.EditorServices.Utility;
+using Xunit;
 
 namespace Microsoft.PowerShell.EditorServices.Test.Session
 {
@@ -40,30 +38,36 @@ namespace Microsoft.PowerShell.EditorServices.Test.Session
         public void CanDetermineIsPathInMemory()
         {
             var tempDir = Environment.GetEnvironmentVariable("TEMP");
-
             var shortDirPath = Path.Combine(tempDir, "GitHub", "PowerShellEditorServices");
             var shortFilePath = Path.Combine(shortDirPath, "foo.ps1");
-            var shortDirUriForm = new Uri(shortDirPath).ToString();
-            var shortFileUriForm = new Uri(shortFilePath).ToString();
-
-            // Test short file absolute paths
-            Assert.False(Workspace.IsPathInMemory(shortDirPath));
-            Assert.False(Workspace.IsPathInMemory(shortFilePath));
-            Assert.False(Workspace.IsPathInMemory(shortDirUriForm));
-            Assert.False(Workspace.IsPathInMemory(shortFileUriForm));
-
-            // Test short file relative paths - not sure we'll ever get these but just in case
-            Assert.False(Workspace.IsPathInMemory("foo.ps1"));
-            Assert.False(Workspace.IsPathInMemory(".." + Path.DirectorySeparatorChar + "foo.ps1"));
-
-            // Test short non-file paths
-            Assert.True(Workspace.IsPathInMemory("untitled:untitled-1"));
             var shortUriForm = "git:/c%3A/Users/Keith/GitHub/dahlbyk/posh-git/src/PoshGitTypes.ps1?%7B%22path%22%3A%22c%3A%5C%5CUsers%5C%5CKeith%5C%5CGitHub%5C%5Cdahlbyk%5C%5Cposh-git%5C%5Csrc%5C%5CPoshGitTypes.ps1%22%2C%22ref%22%3A%22~%22%7D";
-            Assert.True(Workspace.IsPathInMemory(shortUriForm));
-
-            // Test long non-file path - known to have crashed PSES
             var longUriForm = "gitlens-git:c%3A%5CUsers%5CKeith%5CGitHub%5Cdahlbyk%5Cposh-git%5Csrc%5CPoshGitTypes%3Ae0022701.ps1?%7B%22fileName%22%3A%22src%2FPoshGitTypes.ps1%22%2C%22repoPath%22%3A%22c%3A%2FUsers%2FKeith%2FGitHub%2Fdahlbyk%2Fposh-git%22%2C%22sha%22%3A%22e0022701fa12e0bc22d0458673d6443c942b974a%22%7D";
-            Assert.True(Workspace.IsPathInMemory(longUriForm));
+
+            var testCases = new[] {
+                // Test short file absolute paths
+                new { IsInMemory = false, Path = shortDirPath },
+                new { IsInMemory = false, Path = shortFilePath },
+                new { IsInMemory = false, Path = new Uri(shortDirPath).ToString() },
+                new { IsInMemory = false, Path = new Uri(shortFilePath).ToString() },
+
+                // Test short file relative paths - not sure we'll ever get these but just in case
+                new { IsInMemory = false, Path = "foo.ps1" },
+                new { IsInMemory = false, Path = ".." + Path.DirectorySeparatorChar + "foo.ps1" },
+
+                // Test short non-file paths
+                new { IsInMemory = true,  Path = "untitled:untitled-1" },
+                new { IsInMemory = false,  Path = shortUriForm },
+
+                // Test long non-file path - known to have crashed PSES
+                new { IsInMemory = true,  Path = longUriForm },
+            };
+
+            foreach (var testCase in testCases)
+            {
+                Assert.True(
+                    Workspace.IsPathInMemory(testCase.Path) == testCase.IsInMemory,
+                    $"Testing path {testCase.Path}");
+            }
         }
     }
 }

--- a/test/PowerShellEditorServices.Test/Session/WorkspaceTests.cs
+++ b/test/PowerShellEditorServices.Test/Session/WorkspaceTests.cs
@@ -56,7 +56,7 @@ namespace Microsoft.PowerShell.EditorServices.Test.Session
 
                 // Test short non-file paths
                 new { IsInMemory = true,  Path = "untitled:untitled-1" },
-                new { IsInMemory = false,  Path = shortUriForm },
+                new { IsInMemory = true,  Path = shortUriForm },
 
                 // Test long non-file path - known to have crashed PSES
                 new { IsInMemory = true,  Path = longUriForm },

